src/rust/ide/src/model/synchronized/execution_context.rs

```diff
@@ -0,0 +1,230 @@
+//! A ExecutionContext model which is synchronized with LanguageServer state.
+
+use crate::prelude::*;
+
+use crate::double_representation::definition::DefinitionName;
+use crate::model::execution_context::StackItem;
+
+use enso_protocol::language_server;
+use json_rpc::error::RpcError;
+
+
+
+// ==============
+// === Errors ===
+// ==============
+
+/// Error returned when module path is invalid, i.e. cannot obtain module name from it.
+#[derive(Clone,Copy,Debug,Fail)]
+#[fail(display="Invalid module path.")]
+pub struct InvalidModulePath {}
+
+
+
+// ==========================
+// === Synchronized Model ===
+// ==========================
+
+/// An ExecutionContext model synchronized with LanguageServer. It will be automatically removed
+/// from LS once dropped.
+#[derive(Debug)]
+pub struct ExecutionContext {
+    id              : model::execution_context::Id,
+    model           : model::ExecutionContext,
+    module_path     : Rc<controller::module::Path>,
+    language_server : Rc<language_server::Connection>,
+    logger          : Logger,
+}
+
+impl ExecutionContext {
+    /// Create new ExecutionContext. It will be created in LanguageServer and the ExplicitCall
+    /// stack frame will be pushed.
+    pub async fn create
+    ( language_server : Rc<language_server::Connection>
+    , module_path     : Rc<controller::module::Path>
+    , root_definition : DefinitionName
+    ) -> FallibleResult<Self> {
+        let logger = Logger::new("ExecutionContext");
+        let model  = model::ExecutionContext::new(root_definition);
+        trace!(logger,"Creating Execution Context.");
+        let id = language_server.client.create_execution_context().await?.context_id;
+        trace!(logger,"Execution Context created. Id:{id}");
+        let this  = Self {id,module_path,model,language_server,logger};
+        this.push_root_frame().await?;
+        trace!(this.logger,"Pushed root frame");
+        Ok(this)
+    }
+
+    async fn push_root_frame(&self) -> FallibleResult<()> {
+        let module_name = controller::module::obtain_module_name_from_path(&*self.module_path);
+        let method_pointer = language_server::MethodPointer {
+            file            : self.module_path.deref().clone(),
+            defined_on_type : module_name.ok_or(InvalidModulePath{})?.to_string(),
+            name            : self.model.root_definition.name.item.clone(),
+        };
+        let this_argument_expression         = default();
+        let positional_arguments_expressions = default();
+        let call = language_server::ExplicitCall {method_pointer,this_argument_expression,
+            positional_arguments_expressions};
+        let frame = language_server::StackItem::ExplicitCall(call);
+        self.language_server.execution_context_push(self.id,frame).await?;
+        Ok(())
```

Comment:
Wont this compile if we just leave  `self.language_server.execution_context_push(self.id,frame).await` as the last line?